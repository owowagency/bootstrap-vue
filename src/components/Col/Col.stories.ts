import Col from '.';
import Container from '@/components/Container';
import Row from '@/components/Row';
import {breakpoints} from '@/composables/useBreakpoints';

const colArg = {control: {type: 'number'}};

export default {
    title: 'Layout/Col',
    component: Col,
    argTypes: {
        cols: colArg,
        ...Object.fromEntries(breakpoints.map(b => [b, colArg])),
        offset: colArg,
        ...Object.fromEntries(breakpoints.map(b => [`offset-${b}`, colArg])),
        order: colArg,
        ...Object.fromEntries(breakpoints.map(b => [`order-${b}`, colArg])),
    },
};

export const Default = (args) => ({
    components: {Col, Container, Row},
    setup: () => ({args}),
    template: `
        <Container>
            <Row
                class="p-3 border"
            >
                <Col
                    class="border bg-light"
                    v-bind="args"
                >
                    I am a column
                </Col>
            </Row>
        </Container>
    `,
});
