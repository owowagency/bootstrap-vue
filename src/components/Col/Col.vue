<template>
    <div :class="classes">
        <!-- @slot Displays content inside the col -->
        <slot />
    </div>
</template>

<script lang="ts">
import {computed, toRef} from 'vue';
import useBreakpoints, {Breakpoint, breakpointProps} from '@/composables/useBreakpoints';
import {colProp} from '@/composables/useCol';
import useClasses from '@/composables/useClasses';
import useStringTemplate from '@/composables/useStringTemplate';

const propsCol = breakpointProps(colProp) as Record<Breakpoint, typeof colProp>;

const propsOffset = breakpointProps(colProp, 'offset-{0}') as Record<`offset${Capitalize<Breakpoint>}`, typeof colProp>;

const propsOrder = breakpointProps(colProp, 'order-{0}') as Record<`order${Capitalize<Breakpoint>}`, typeof colProp>;
</script>

<script lang="ts" setup>
const props = defineProps({
    /**
     * The amount of cols the column should span. Results in: `.col-*`
     */
    cols: colProp,
    ...propsCol,
    /**
     * The amount of cols the column should be offset. Results in `.offset-*`
     */
    offset: colProp,
    ...propsOffset,
    /**
     * The position in order of the column. Results in `.order-*`
     */
    order: colProp,
    ...propsOrder,
});

const colProps = {
    sm: toRef(props, 'sm'),
    md: toRef(props, 'md'),
    lg: toRef(props, 'lg'),
    xl: toRef(props, 'xl'),
    xxl: toRef(props, 'xxl'),
};

const colClass = computed(() => {
    // Add the `.col` class by default when no cols have been specified.
    if (props.cols) {
        return useStringTemplate('col-{0}', props.cols).templatedString.value;
    }

    return 'col';
});

const colClasses = computed(() => {
    // When the property is empty string it would result in for example
    // `col-md-`. These classes should be transformed to `col-md`.
    return useBreakpoints(colProps, 'col-{0}-{1}')
        .breakpointClasses
        .value
        .map(c => c.replace(/\-$/, ''));
});

const offsetProps = {
    sm: toRef(props, 'offsetSm'),
    md: toRef(props, 'offsetMd'),
    lg: toRef(props, 'offsetLg'),
    xl: toRef(props, 'offsetXl'),
    xxl: toRef(props, 'offsetXxl'),
};

const offsetClass = computed(() => useStringTemplate('offset-{0}', props.offset).templatedString.value);

const offsetClasses = useBreakpoints(offsetProps, 'offset-{0}-{1}').breakpointClasses;

const orderProps = {
    sm: toRef(props, 'orderSm'),
    md: toRef(props, 'orderMd'),
    lg: toRef(props, 'orderLg'),
    xl: toRef(props, 'orderXl'),
    xxl: toRef(props, 'orderXxl'),
};

const orderClass = computed(() => useStringTemplate('order-{0}', props.order).templatedString.value);

const orderClasses = useBreakpoints(orderProps, 'order-{0}-{1}').breakpointClasses;

const {classes} = useClasses(computed(() => [
    colClass.value,
    ...colClasses.value,
    offsetClass.value,
    ...offsetClasses.value,
    orderClass.value,
    ...orderClasses.value,
]));
</script>

<docs>
```vue
<Col
    cols="10"
    sm="8"
    md="6"
    offset="1"
    offset-sm="2"
    offset-md="3"
>
    I am a column
</Col>
```

## Imported Props

| Prop name    | Description                                                                   | Type            | Values | Default   |
| ------------ | ----------------------------------------------------------------------------- | --------------- | ------ | --------- |
| cols[size]   | The amount of cols the column should span. Results in `.col-[size]-*`         | `int`\|`string` | -      | undefined |
| offset[Size] | The amount of cols the column should be offset. Results in `.offset-[size]-*` | `int`\|`string` | -      | undefined |
| order[Size]  | The position in order of the column. Results in `.order-[size]-*`             | `int`\|`string` | -      | undefined |
</docs>
