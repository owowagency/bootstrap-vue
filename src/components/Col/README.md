# Col

```vue
<Col cols="10" sm="8" md="6" offset="1" offset-sm="2" offset-md="3">
    I am a column
</Col>
```

## Imported Props

| Prop name    | Description                                                                   | Type            | Values | Default   |
| ------------ | ----------------------------------------------------------------------------- | --------------- | ------ | --------- |
| cols[size]   | The amount of cols the column should span. Results in `.col-[size]-*`         | `int`\|`string` | -      | undefined |
| offset[Size] | The amount of cols the column should be offset. Results in `.offset-[size]-*` | `int`\|`string` | -      | undefined |
| order[Size]  | The position in order of the column. Results in `.order-[size]-*`             | `int`\|`string` | -      | undefined |

## Props

| Prop name | Description                                                            | Type    | Values | Default |
| --------- | ---------------------------------------------------------------------- | ------- | ------ | ------- |
| cols      | The amount of cols the column should span. Results in: `.col-*`        | colProp | -      |         |
| offset    | The amount of cols the column should be offset. Results in `.offset-*` | colProp | -      |         |
| order     | The position in order of the column. Results in `.order-*`             | colProp | -      |         |

## Slots

| Name    | Description                     | Bindings |
| ------- | ------------------------------- | -------- |
| default | Displays content inside the col |          |
