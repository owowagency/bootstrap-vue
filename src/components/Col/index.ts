import Col from './Col.vue';

export default Col;
