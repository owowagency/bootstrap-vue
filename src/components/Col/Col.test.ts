import Col from '.';
import {describe} from 'vitest';

describe('template', () => {
    componentRenderTest(Col);

    componentSlotRenderTest(Col);

    componentWrapperClassTest(Col, {cols: 1}, 'col-1');
    componentWrapperClassTest(Col, {sm: 2}, 'col-sm-2');
    componentWrapperClassTest(Col, {md: 3}, 'col-md-3');
    componentWrapperClassTest(Col, {lg: 4}, 'col-lg-4');
    componentWrapperClassTest(Col, {xl: 5}, 'col-xl-5');
    componentWrapperClassTest(Col, {xxl: 6}, 'col-xxl-6');

    componentWrapperClassTest(Col, {offset: 1}, 'offset-1');
    componentWrapperClassTest(Col, {offsetSm: 2}, 'offset-sm-2');
    componentWrapperClassTest(Col, {offsetMd: 3}, 'offset-md-3');
    componentWrapperClassTest(Col, {offsetLg: 4}, 'offset-lg-4');
    componentWrapperClassTest(Col, {offsetXl: 5}, 'offset-xl-5');
    componentWrapperClassTest(Col, {offsetXxl: 6}, 'offset-xxl-6');

    componentWrapperClassTest(Col, {order: 1}, 'order-1');
    componentWrapperClassTest(Col, {orderSm: 2}, 'order-sm-2');
    componentWrapperClassTest(Col, {orderMd: 3}, 'order-md-3');
    componentWrapperClassTest(Col, {orderLg: 4}, 'order-lg-4');
    componentWrapperClassTest(Col, {orderXl: 5}, 'order-xl-5');
    componentWrapperClassTest(Col, {orderXxl: 6}, 'order-xxl-6');
});
