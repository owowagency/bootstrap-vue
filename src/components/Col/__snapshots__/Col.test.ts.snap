// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div class=\\"col\\">
  <!-- @slot Displays content inside the col -->
</div>"
`;
